devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+    expiry: Option<Instant>,
+}
+
+impl<S> VsockChannel for VsockConnection<S>
+where
+    S: Read + Write + AsRawFd,
+{
+    /// Fill in a vsock packet, to be delivered to our peer (the guest driver).
+    ///
+    /// As per the `VsockChannel` trait, this should only be called when there is data to be fetched
+    /// from the channel (i.e. `has_pending_rx()` is true). Otherwise, it will error out with
+    /// `VsockError::NoData`.
+    /// Pending RX indications are set by other mutable actions performed on the channel. For instance,
+    /// `send_pkt()` could set an Rst indication, if called with a VSOCK_OP_SHUTDOWN packet, or
+    /// `notify()` could set a Rw indication (a data packet can be fetched from the channel), if data
+    /// was ready to be read from the host stream.
+    ///
+    /// Returns:
+    /// - `Ok(())`: the packet has been successfully filled in and is ready for delivery;
+    /// - `Err(VsockError::NoData)`: there was no data available with which to fill in the packet;
+    /// - `Err(VsockError::PktBufMissing)`: the packet would've been filled in with data, but it is
+    ///   missing the data buffer.
+    ///
+    fn recv_pkt(&mut self, pkt: &mut VsockPacket) -> VsockResult<()> {
+        // Perform some generic initialization that is the same for any packet operation (e.g.
+        // source, destination, credit, etc).
+        self.init_pkt(pkt);
+
+        // If forcefull termination is pending, there's no point in checking for anything else.
+        // It's dead, Jim.
+        if self.pending_rx.remove(PendingRx::Rst) {
+            pkt.set_op(uapi::VSOCK_OP_RST);
+            return Ok(());
+        }
+
+        // Next up: if we're due a connection confirmation, that's all we need to know to fill in
+        // this packet.
+        if self.pending_rx.remove(PendingRx::Response) {
+            self.state = ConnState::Established;
+            pkt.set_op(uapi::VSOCK_OP_RESPONSE);
+            return Ok(());
+        }
+
+        // Same thing goes for locally-initiated connections, that need to yield a connection request.
+        if self.pending_rx.remove(PendingRx::Request) {
+            self.expiry =
+                Some(Instant::now() + Duration::from_millis(defs::CONN_REQUEST_TIMEOUT_MS));
+            pkt.set_op(uapi::VSOCK_OP_REQUEST);
+            return Ok(());
+        }
+
+        // A credit update is basically a no-op, so we should only waste a perfectly fine RX buffer
+        // on it if we really have nothing else to say.
+        if self.pending_rx.remove(PendingRx::CreditUpdate) && !self.has_pending_rx() {
+            pkt.set_op(uapi::VSOCK_OP_CREDIT_UPDATE);
+            self.last_fwd_cnt_to_peer = self.fwd_cnt;
+            return Ok(());
+        }
+
+        // Alright, if we got to here, we need to cough up a data packet. We've already checked for
+        // all other pending RX indications.
+        self.pending_rx.remove(PendingRx::Rw);
```

Comment:
I was not referring to calling `recv_pkt` being an invalid state. Let me rephrase it.
`self.pending_rx.remove` returns a bool (that you ignore), right? If the function were to return false, is there sense in going ahead and reading data?

Reply:
Ah, right. I've gone through this section again and simplified some stuff. Now, a read is not attempted if `self.pending_rx` doesn't contain `PendingRx::Rw`. That's because reading (and writing, for that matter) data is fully driven by epoll. So, we shouldn't be attempting to read before `self.notify()` is called with `EPOLLIN` (which sets `Rw` inside `self.pendingRx`).